homeassistant/components/control4/media_player.py

```diff
@@ -0,0 +1,407 @@
+"""Platform for Control4 Rooms Media Players."""
+from __future__ import annotations
+
+from datetime import timedelta
+import enum
+import logging
+from typing import Any
+
+from pyControl4.error_handling import C4Exception
+from pyControl4.room import C4Room
+
+from homeassistant.components.media_player import (
+    MediaPlayerDeviceClass,
+    MediaPlayerEntity,
+    MediaPlayerEntityFeature,
+    MediaPlayerState,
+    MediaType,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_SCAN_INTERVAL
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from . import Control4Entity
+from .const import CONF_DIRECTOR, CONF_DIRECTOR_ALL_ITEMS, CONF_UI_CONFIGURATION, DOMAIN
+from .director_utils import update_variables_for_config_entry
+
+_LOGGER = logging.getLogger(__name__)
+
+CONTROL4_POWER_STATE = "POWER_STATE"
+CONTROL4_VOLUME_STATE = "CURRENT_VOLUME"
+CONTROL4_MUTED_STATE = "IS_MUTED"
+CONTROL4_CURRENT_VIDEO_DEVICE = "CURRENT_VIDEO_DEVICE"
+CONTROL4_PLAYING = "PLAYING"
+CONTROL4_PAUSED = "PAUSED"
+CONTROL4_STOPPED = "STOPPED"
+CONTROL4_MEDIA_INFO = "CURRENT MEDIA INFO"
+
+CONTROL4_PARENT_ID = "parentId"
+
+VARIABLES_OF_INTEREST = {
+    CONTROL4_POWER_STATE,
+    CONTROL4_VOLUME_STATE,
+    CONTROL4_MUTED_STATE,
+    CONTROL4_CURRENT_VIDEO_DEVICE,
+    CONTROL4_MEDIA_INFO,
+    CONTROL4_PLAYING,
+    CONTROL4_PAUSED,
+    CONTROL4_STOPPED,
+}
+
+
+class _SourceType(enum.Enum):
+    AUDIO = 1
+    VIDEO = 2
+
+
+class _RoomSource:
+    """Room Source Data."""
+
+    def __init__(self, source_type: set[_SourceType], idx: int, name: str) -> None:
+        """Initialize Room Source Data."""
+        self.source_type = source_type
+        self.idx = idx
+        self.name = name
```

Comment:
I would make this a dataclass, but that can be done in a follow-up PR
```suggestion
@dataclass
class _RoomSource:
    """Room Source Data."""

    source_type: set[_SourceType]
    idx: int
    name: str
```

Reply:
Done